src/Build/BackEnd/Components/RequestBuilder/IntrinsicTasks/ItemGroupLoggingHelper.cs

```diff
@@ -93,55 +91,92 @@ internal static string GetParameterText(string prefix, string parameterName, ILi
+        {
+            if (parameterValue is string text)
             {
-                return (string)Convert.ChangeType(parameterValue, typeof(string), CultureInfo.CurrentCulture);
+                sb.Append(text);
             }
-            else if (typeof(ITaskItem).GetTypeInfo().IsAssignableFrom(type.GetTypeInfo()))
+            else if (parameterValue is ITaskItem item)
             {
-                var item = ((ITaskItem)parameterValue);
-                string result = item.ItemSpec;
-
-                var customMetadata = item.CloneCustomMetadata();
+                sb.Append(item.ItemSpec);
 
-                if (customMetadata.Count > 0)
+                if (includeMetadata)
                 {
-                    result += "\n";
-                    var names = new List<string>();
+                    var customMetadata = item.CloneCustomMetadata();
+                    int count = customMetadata.Count;
 
-                    foreach (string name in customMetadata.Keys)
+                    if (count > 0)
                     {
-                        names.Add(name);
-                    }
+                        sb.Append('\n');
+
+                        // need to initialize the thread static on each new thread
+                        if (keyValuePairList == null)
+                        {
+                            keyValuePairList = new List<KeyValuePair<string, string>>(count);
+                        }
 
-                    names.Sort();
+                        var customMetadataDictionary = customMetadata as IDictionary<string, string>;
+                        if (customMetadataDictionary != null)
+                        {
+                            foreach (KeyValuePair<string, string> kvp in customMetadataDictionary)
+                            {
+                                keyValuePairList.Add(kvp);
+                            }
+                        }
+                        else
+                        {
+                            foreach (DictionaryEntry kvp in customMetadata)
+                            {
+                                keyValuePairList.Add(new KeyValuePair<string, string>((string)kvp.Key, (string)kvp.Value));
+                            }
+                        }
 
-                    for (int i = 0; i < names.Count; i++)
-                    {
-                        result += "                " + names[i] + "=" + customMetadata[names[i]];
+                        if (count > 1)
+                        {
+                            keyValuePairList.Sort((l, r) => StringComparer.OrdinalIgnoreCase.Compare(l.Key, r.Key));
+                        }
 
-                        if (i < names.Count - 1)
+                        for (int i = 0; i < count; i++)
                         {
-                            result += "\n";
+                            var kvp = keyValuePairList[i];
+                            sb.Append("                ");
```

Comment:
Can this string be allocated outside the function so we don't allocate it multiple times?

Reply:
These aren't allocated. It's just `ldstr` which gets you a pointer to the Strings metadata heap cheaply: https://sharplab.io/#v2:EYLgZgpghgLgrgJwgZwLRIMaOQSwG4SoAOsMECAdsgDQwg4A2APgAIBMAjALABQLAzAAJ2ggMKCA3r0EzhQlgBZBAWQAUASknTZOlhwCcqgERgA9qaPqA3NpkBfXnaA=